flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java

```diff
@@ -195,12 +231,15 @@ public void processWatermark(Watermark mark) throws Exception {
 
     @Override
     public void endInput() throws Exception {
-        endOfInput = true;
-        sinkWriter.flush(true);
-        emitCommittables(Long.MAX_VALUE);
+        if (!endOfInput) {
+            endOfInput = true;
```

Comment:
Would it make sense to disallow `SinkWriterOperator` to `processElement` anymore after this?

Reply:
Interesting idea but wouldn't we validate the framework then?

Or would you also add the assertion after recovery? That's kinda done with the `ClosedSinkWriter`.

Reply:
I added the assertion. It can't hurt because we may have upstream operators that misbehave (custom operators).